### PR TITLE
distilbert-flax

### PR DESCRIPTION
**Pull request by kamalkraj**
# What does this PR do?
DistilBert Flax

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@VictorSanh @patrickvonplaten

### CODE REVIEW

File: src/transformers/models/distilbert/modeling_flax_distilbert.py [hunk 1]
```diff
@@ -0,0 +1,928 @@
...
+
+
+logger = logging.get_logger(__name__)
+
+_CHECKPOINT_FOR_DOC = "distilbert-base-uncased"
+_CONFIG_FOR_DOC = "DistilBertConfig"
+_TOKENIZER_FOR_DOC = "DistilBertTokenizer"
+
+
+DISTILBERT_START_DOCSTRING = r"""
```
Comment (patrickvonplaten) [inline src/transformers/models/distilbert/modeling_flax_distilbert.py:48]:
```suggestion
FLAX_DISTILBERT_START_DOCSTRING = r"""
```

And Ideally all the following docstrings :-)

File: src/transformers/models/distilbert/modeling_flax_distilbert.py [hunk 2]
```diff
@@ -0,0 +1,928 @@
...
+
+    This model inherits from :class:`~transformers.PreTrainedModel`. Check the superclass documentation for the generic
```
Comment (patrickvonplaten) [inline src/transformers/models/distilbert/modeling_flax_distilbert.py:50]: This text is for PyTorch - can we maybe copy paste it from `FlaxBert` and adapt it slightly? :-)

File: src/transformers/models/distilbert/modeling_flax_distilbert.py [hunk 3]
```diff
@@ -0,0 +1,928 @@
...
+    Parameters:
+        config (:class:`~transformers.DistilBertConfig`): Model configuration class with all the parameters of the model.
+            Initializing with a config file does not load the weights associated with the model, only the
+            configuration. Check out the :meth:`~transformers.PreTrainedModel.from_pretrained` method to load the model
+            weights.
+"""
+
+DISTILBERT_INPUTS_DOCSTRING = r"""
+    Args:
+        input_ids (:obj:`torch.LongTensor` of shape :obj:`({0})`):
```
Comment (patrickvonplaten) [inline src/transformers/models/distilbert/modeling_flax_distilbert.py:67]:
```suggestion
        input_ids (:obj:`jnp.ndarray` of shape :obj:`({0})`):
```

File: src/transformers/models/distilbert/modeling_flax_distilbert.py [hunk 4]
```diff
@@ -0,0 +1,928 @@
...
+    dtype: jnp.dtype = jnp.float32  # the dtype of the computation
+
+    def setup(self):
+        self.layers = [
+            FlaxTransformerBlock(self.config, name=str(i), dtype=self.dtype) for i in range(self.config.n_layers)
+        ]
+
+    def __call__(
+        self,
+        x,
```
Comment (patrickvonplaten) [inline src/transformers/models/distilbert/modeling_flax_distilbert.py:349]: (nit) - can we maybe replace all `x` with `hidden_states`?

File: src/transformers/models/distilbert/modeling_flax_distilbert.py [hunk 5]
```diff
@@ -0,0 +1,928 @@
...
+        return_dict: bool = True,
+    ):
+        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+        return_dict = return_dict if return_dict is not None else self.config.return_dict
+
+        if head_mask is not None:
+            raise NotImplementedError
```
Comment (patrickvonplaten) [inline src/transformers/models/distilbert/modeling_flax_distilbert.py:530]: I don't fully understand this here - the model should work even if the `head_mask` is not supplied no?

File: src/transformers/models/distilbert/modeling_flax_distilbert.py [hunk 6]
```diff
@@ -0,0 +1,928 @@
...
+        else:
+            head_mask = [None] * self.config.num_hidden_layers
```
Comment (patrickvonplaten) [inline src/transformers/models/distilbert/modeling_flax_distilbert.py:532]: Can we just delete `head_mask` all together? :-)

File: tests/test_modeling_flax_distilbert.py [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+            model = model_class_name.from_pretrained("distilbert-base-uncased", from_pt=True)
+            outputs = model(np.ones((1, 1)))
+            self.assertIsNotNone(outputs)
+
+
+@require_flax
+class DistilBertModelIntegrationTest(unittest.TestCase):
+    @slow
+    def test_inference_no_head_absolute_embedding(self):
+        model = FlaxDistilBertModel.from_pretrained("distilbert-base-uncased", from_pt=True)
```
Comment (patrickvonplaten) [inline tests/test_modeling_flax_distilbert.py:144]:
```suggestion
        model = FlaxDistilBertModel.from_pretrained("distilbert-base-uncased")
```

Comment (patrickvonplaten) [reply tests/test_modeling_flax_distilbert.py:144]: Just added the flax model weights :-)

File: tests/test_modeling_flax_distilbert.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    @slow
+    def test_model_from_pretrained(self):
+        for model_class_name in self.all_model_classes:
+            model = model_class_name.from_pretrained("distilbert-base-uncased", from_pt=True)
+            outputs = model(np.ones((1, 1)))
+            self.assertIsNotNone(outputs)
+
+
+@require_flax
+class DistilBertModelIntegrationTest(unittest.TestCase):
```
Comment (patrickvonplaten) [inline tests/test_modeling_flax_distilbert.py:141]:
```suggestion
class FlaxDistilBertModelIntegrationTest(unittest.TestCase):
```

File: tests/test_modeling_flax_distilbert.py [hunk 9]
```diff
@@ -0,0 +1,152 @@
+# Copyright 2020 The HuggingFace Team. All rights reserved.
```
Comment (patrickvonplaten) [inline tests/test_modeling_flax_distilbert.py:1]:
```suggestion
# Copyright 2021 The HuggingFace Team. All rights reserved.
```

**Comment by patrickvonplaten** [comment]
Great Great job @kamalkraj! 

Think the only major thing to update is to docs in the modeling file (at the moment it looks like it's the PyTorch docs, but should be Flax :-))

File: src/transformers/models/distilbert/modeling_flax_distilbert.py (hunk 1 above)
Comment (kamalkraj) [reply src/transformers/models/distilbert/modeling_flax_distilbert.py]: done.
https://github.com/huggingface/transformers/pull/13324/commits/7550233f6eb7610390c6002134ab5cd77baf5a53

File: src/transformers/models/distilbert/modeling_flax_distilbert.py (hunk 3 above)
Comment (kamalkraj) [reply src/transformers/models/distilbert/modeling_flax_distilbert.py]: done
https://github.com/huggingface/transformers/pull/13324/commits/7550233f6eb7610390c6002134ab5cd77baf5a53

File: src/transformers/models/distilbert/modeling_flax_distilbert.py (hunk 2 above)
Comment (kamalkraj) [reply src/transformers/models/distilbert/modeling_flax_distilbert.py]: done
https://github.com/huggingface/transformers/pull/13324/commits/7550233f6eb7610390c6002134ab5cd77baf5a53

File: src/transformers/models/distilbert/modeling_flax_distilbert.py (hunk 4 above)
Comment (kamalkraj) [reply src/transformers/models/distilbert/modeling_flax_distilbert.py]: done.
https://github.com/huggingface/transformers/pull/13324/commits/f972d2d72710dd16822be42ba0f9b1503410f3aa

File: src/transformers/models/distilbert/modeling_flax_distilbert.py (hunk 6 above)
Comment (kamalkraj) [reply src/transformers/models/distilbert/modeling_flax_distilbert.py]: done
https://github.com/huggingface/transformers/pull/13324/commits/a703881fcb1f18ccbe3f7a06dfa52010594b4bd7

File: tests/test_modeling_flax_distilbert.py (hunk 7 above)
Comment (kamalkraj) [reply tests/test_modeling_flax_distilbert.py]: done.
https://github.com/huggingface/transformers/pull/13324/commits/004e1bcb3c9585444034c0c665bf45c3b3fbf0c5

File: tests/test_modeling_flax_distilbert.py (hunk 8 above)
Comment (kamalkraj) [reply tests/test_modeling_flax_distilbert.py]: done.
https://github.com/huggingface/transformers/pull/13324/commits/004e1bcb3c9585444034c0c665bf45c3b3fbf0c5

File: tests/test_modeling_flax_distilbert.py (hunk 9 above)
Comment (kamalkraj) [reply tests/test_modeling_flax_distilbert.py]: done.
https://github.com/huggingface/transformers/pull/13324/commits/32ac65f6761c4085c9b3ca00b5548cb41533555e

**Comment by kamalkraj** [comment]
@patrickvonplaten 
Thanks for the review.
Done changes according to your review.

**Comment by stefan-it** [comment]
Hi @kamalkraj , I'm also really interested in that PR - thanks for adding it :hugs: 

Do you also plan tu add a script for the distillation process (like it is done in the ["old" script](https://github.com/huggingface/transformers/tree/master/examples/research_projects/distillation)), as I would like to re-distillate some of my previous DistilBERT models (I don't have access to multi GPU setups, only to TPUs at the moment).

**Comment by kamalkraj** [comment]
Hi @stefan-it,

I will go through the scripts and pings you. 
I have multi-GPU access. Which TPU do you use? v3-8 ?

**Comment by patrickvonplaten** [comment]
```
JAX_PLATFORM_NAME=cpu RUN_SLOW=1 pytest tests/test_modeling_flax_distilbert.py::FlaxDistilBertModelIntegrationTest::test_inference_no_head_absolute_embedding
```

passes and the code looks good :-) Ready to merge IMO :tada: ! 

@patil-suraj the slow test doesn't pass on TPU since distilbert has pretty extreme activations in the forward pass like a couple of other models. We need to think a bit how to adapt the slow test depending on whether they're run on TPU or not in general...

**Comment by patrickvonplaten** [comment]
Great work @kamalkraj !

**Pull request merged**